### PR TITLE
feat(platform-wallet): expose address funding fee estimate

### PR DESCRIPTION
**Pull request by llbartekll**
## Issue being fixed or feature implemented

Adds a small static address-funding fee estimate surface so clients can reserve the DPP minimum-required fee without using the abandoned DAPI quote engine.

This targets `v4.2-dev` directly because #4501 has been merged. It intentionally does not estimate state-aware metered GroveDB cost or query live network state. It mirrors `AddressFundingFromAssetLockTransition::calculate_min_required_fee` from DPP and exposes that count-based formula through FFI and Swift.

## What was done?

- Added `calculate_address_funding_from_asset_lock_min_required_fee(input_count, output_count, platform_version)` in DPP and kept the existing transition trait implementation delegated to it.
- Added rustdoc on the public DPP helper.
- Added table-driven DPP test vectors for 0/1/2 outputs, mixed input/output counts, and a larger 10-input/100-output case.
- Added a DPP test proving the count-based helper matches `AddressFundingFromAssetLockTransition::calculate_min_required_fee` for real transition values.
- Added `platform_wallet_address_funding_estimate_fee` in `rs-platform-wallet-ffi`.
- Added Swift SDK wrapper `PlatformWalletManager.estimateAddressFundingFee(inputCount:outputCount:)`.
- Pinned the current 0-input / 2-output reserve used by Core -> Platform topups at `62_000_000` credits.

## How has this been tested?

- `cargo test -p dpp count_based_helper_matches`
- `CARGO_INCREMENTAL=0 CARGO_PROFILE_TEST_DEBUG=0 CARGO_PROFILE_DEV_DEBUG=0 cargo test -p platform-wallet-ffi platform_addresses::funding_fee`
- `./build_ios.sh --target sim`
- `git diff --check`

## Notes

This replaces the closed fee-quote PR stack (#4444, #4445, #4446, #4447). The iOS app can use this static reserve together with the sender-owned remainder semantics merged in #4501: explicit recipient amount stays exact, the real Platform fee is deducted from the sender remainder output, and any unused reserve remains in the sender Platform balance.

Live-network testing belongs to the app/topup flow, not this DPP helper: this helper is deliberately a protocol-versioned minimum reserve formula, not an assertion that live GroveDB metered cost equals this value.

## Summary by CodeRabbit

* **New Features**
  * Added local address-funding fee estimation to the platform wallet.
  * Swift SDK users can estimate required fees using input and output counts without a network request.
  * Exposed fee estimation through platform wallet bindings.

* **Bug Fixes**
  * Enforced minimum funding fees before asset-lock funding is signed or broadcast.
  * Improved handling of insufficient balances and drain transactions.

* **Tests**
  * Added coverage for fee calculations, validation, and funding thresholds.
  * Improved certificate renewal diagnostics and document identifier comparisons.

### CODE REVIEW

File: packages/rs-dpp/src/state_transition/state_transitions/address_funds/address_funding_from_asset_lock_transition/state_transition_estimated_fee_validation.rs [hunk 1]
```diff
@@ -8,30 +8,59 @@ use crate::state_transition::{
 use crate::ProtocolError;
 use platform_version::version::PlatformVersion;
 
+pub fn calculate_address_funding_from_asset_lock_min_required_fee(
```
Comment (lklimek) [inline packages/rs-dpp/src/state_transition/state_transitions/address_funds/address_funding_from_asset_lock_transition/state_transition_estimated_fee_validation.rs:16]: rustdoc please, like 3 lines - it's a public  API

File: packages/rs-dpp/src/state_transition/state_transitions/address_funds/address_funding_from_asset_lock_transition/state_transition_estimated_fee_validation.rs [hunk 2]
```diff
@@ -8,30 +8,59 @@ use crate::state_transition::{
...
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::version::LATEST_PLATFORM_VERSION;
+
+    #[test]
+    fn count_based_helper_matches_current_static_address_funding_fee() {
```
Comment (lklimek) [inline packages/rs-dpp/src/state_transition/state_transitions/address_funds/address_funding_from_asset_lock_transition/state_transition_estimated_fee_validation.rs:97]: I would suggest more test vectors for various combinations, edge cases, etc, to prove this really works (tested on the live network to ensure it's enough).

**Review by lklimek — CHANGES_REQUESTED** [review]

File: packages/rs-dpp/src/state_transition/state_transitions/address_funds/address_funding_from_asset_lock_transition/state_transition_estimated_fee_validation.rs (hunk 1 above)
Comment (llbartekll) [reply packages/rs-dpp/src/state_transition/state_transitions/address_funds/address_funding_from_asset_lock_transition/state_transition_estimated_fee_validation.rs:16]: Added rustdoc on the public helper in e6dacf42f8.

File: packages/rs-dpp/src/state_transition/state_transitions/address_funds/address_funding_from_asset_lock_transition/state_transition_estimated_fee_validation.rs (hunk 2 above)
Comment (llbartekll) [reply packages/rs-dpp/src/state_transition/state_transitions/address_funds/address_funding_from_asset_lock_transition/state_transition_estimated_fee_validation.rs:97]: Added table-driven vectors for 0/1/2 outputs, mixed input/output counts, and a larger 10-input/100-output case. Also added a test that builds real `AddressFundingFromAssetLockTransition` values and proves this helper matches `calculate_min_required_fee` for those shapes. I clarified in the PR body that this helper is a protocol-versioned minimum reserve formula, not a live-network state-aware fee quote; live-network validation belongs to the app topup flow.

File: packages/rs-platform-wallet-ffi/src/platform_addresses/funding_fee.rs [hunk 3]
```diff
@@ -0,0 +1,154 @@
+//! FFI helpers for address-funding fee estimates.
+
+use dpp::state_transition::address_funding_from_asset_lock_transition::calculate_address_funding_from_asset_lock_min_required_fee;
+
+use crate::check_ptr;
+use crate::error::*;
+use crate::handle::{Handle, PLATFORM_WALLET_MANAGER_STORAGE};
+
+/// Estimate the static minimum fee reserve (in credits) for an
```
Comment (Claudius-Maginificent) [inline packages/rs-platform-wallet-ffi/src/platform_addresses/funding_fee.rs]: **[SEC-002, blocking] Estimate is denominated in credits while the funding entry point it exists to size takes duffs — 1000x mis-sizing trap**

This rustdoc says the purpose is to let hosts "reserve Core lock value" — a duffs quantity — but the function returns credits. The only in-repo consumer, `platform_address_wallet_fund_from_asset_lock_signer`, takes `amount_duffs: u64`. Both sides are bare `u64` with no unit in the name and no conversion helper, so the natural composition (estimate → fund with the estimate) is wrong by 1000x. The sibling shielded flow solves this correctly in Rust with a ceiling conversion (`pool_fee_credits / CREDITS_PER_DUFF + 1`); this PR exports the raw credits value instead.

**Recommendation:** Either (a) return duffs directly with a ceiling conversion mirroring the shielded flow (rename to `..._estimate_lock_duffs`), or (b) put the unit in both symbol names (`..._fee_credits`) and ship a companion duffs converter. Also fix the "reserve Core lock value" phrasing in this doc comment either way.

File: packages/swift-sdk/Sources/SwiftDashSDK/PlatformWallet/PlatformWalletManagerAddressFunding.swift [hunk 4]
```diff
@@ -0,0 +1,39 @@
+import Foundation
+import DashSDKFFI
+
+extension PlatformWalletManager {
+    /// Static minimum fee reserve (in credits) for an
```
Comment (Claudius-Maginificent) [inline packages/swift-sdk/Sources/SwiftDashSDK/PlatformWallet/PlatformWalletManagerAddressFunding.swift]: **[SEC-002, blocking]** Same credits/duffs unit mismatch as `funding_fee.rs` — see that file's comment for the full writeup. This Swift-facing doc should also spell out the unit (credits) so app developers don't compose this directly into `amount_duffs`.

File: packages/rs-dpp/src/state_transition/state_transitions/address_funds/address_funding_from_asset_lock_transition/state_transition_estimated_fee_validation.rs [hunk 5]
```diff
@@ -8,30 +8,130 @@ use crate::state_transition::{
 use crate::ProtocolError;
 use platform_version::version::PlatformVersion;
 
+/// Calculate the static minimum required fee for address funding from an asset lock.
+///
+/// This is the count-based DPP admission-floor formula used by
+/// `AddressFundingFromAssetLockTransition::calculate_min_required_fee`, exposed for
+/// callers that need to reserve funds before constructing the transition.
+pub fn calculate_address_funding_from_asset_lock_min_required_fee(
```
Comment (Claudius-Maginificent) [inline packages/rs-dpp/src/state_transition/state_transitions/address_funds/address_funding_from_asset_lock_transition/state_transition_estimated_fee_validation.rs]: **[SEC-004] Counts accepted unbounded; saturation returns Success with no overflow signal, unlike the sibling estimate FFI**

Neither this helper nor the FFI export validates `input_count`/`output_count` against `max_address_inputs`/`max_address_outputs` (16/128 on latest version), and the arithmetic is entirely `saturating_*`, so a garbage count (e.g. a non-Swift host passing `(size_t)-1`) silently returns a `u64`-scale value with `Success` instead of an error. The sibling `platform_wallet_shielded_estimate_fee` (funding_fee.rs analog: `shielded_send.rs:230-239`) surfaces this as `ErrorArithmeticOverflow`.

**Recommendation:** Validate counts at the FFI boundary (reject out-of-range with `ErrorInvalidParameter`), and return `Result<Credits, ProtocolError>` using `checked_*` so overflow is representable rather than silent.

File: packages/rs-dpp/src/state_transition/state_transitions/address_funds/address_funding_from_asset_lock_transition/state_transition_estimated_fee_validation.rs [hunk 6]
```diff
@@ -8,30 +8,130 @@ use crate::state_transition::{
 use crate::ProtocolError;
 use platform_version::version::PlatformVersion;
 
+/// Calculate the static minimum required fee for address funding from an asset lock.
```
Comment (Claudius-Maginificent) [inline packages/rs-dpp/src/state_transition/state_transitions/address_funds/address_funding_from_asset_lock_transition/state_transition_estimated_fee_validation.rs]: **[DOC-002] The silent `output_count.max(1)` clamp is documented at none of the three layers**

`output_count.max(1)` silently reprices a 0-output request as 1 output. This is invisible in this rustdoc, the FFI rustdoc, and the Swift doc comment — the FFI test `estimate_fee_preserves_one_output_minimum_for_zero_outputs` proves it was known, it just never reached a caller-visible doc.

**Recommendation:** One sentence in each of the three doc comments: "`output_count` is floored at 1 — the transition always carries at least one output, so 0 and 1 price identically."

**Review by lklimek — APPROVED** [review]
